### PR TITLE
LABS-1053 Make the ingest table an in-memory btree

### PR DESCRIPTION
**Pull request by quchenhao**
Rename the cache_resident configuration to use another flag WT_BTREE_NO_EVICT. We cannot use the cache_resident configuration for the ingest table as it prohibits running eviction on the tree.

Add another option to create an in-memory btree for non-in-memory database. We will create a file for this btree in this case but we never write anything to it. For the in-memory btree, we reconcile its pages the same way as the btrees in in-memory database during eviction. While checkpoint always skips this type of btrees.

Use the in-memory btree for the ingest table. The ingest table cannot write data to the disk as it shouldn't insert updates to the history store.

The reconciliation of the in-memory btree can be used as the garbage collection mechanism for the ingest table.

### CODE REVIEW

File: src/docs/doxygen.log
Comment (korteland) [inline src/docs/doxygen.log:1]: Stray `git add` for this file

File: src/btree/bt_split.c [hunk 1]
```diff
@@ -1450,7 +1451,8 @@ __split_multi_inmem(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *multi, WT
          * them back to their original update chains. Truncate before we restore them to ensure the
          * size of the page is correct.
          */
-        if (supd->onpage_upd != NULL && !F_ISSET(S2C(session), WT_CONN_IN_MEMORY)) {
+        if (supd->onpage_upd != NULL && !F_ISSET(S2C(session), WT_CONN_IN_MEMORY) &&
+          !F_ISSET(S2BT(session), WT_BTREE_IN_MEMORY)) {
```
Comment (korteland) [inline src/btree/bt_split.c:1455]: Nitpick, we're now doing a lot of these `WT_CONN_IN_MEMORY && WT_BTREE_IN_MEMORY` checks paired together. Should we have a macro to combine them?

File: src/evict/evict_lru.c [hunk 2]
```diff
@@ -2096,9 +2100,11 @@ __evict_walk_tree(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, u_int max_ent
         }
 
         /* Skip pages we don't want. */
-        want_page = (F_ISSET(cache, WT_CACHE_EVICT_CLEAN) && !modified) ||
+        want_page = (F_ISSET(cache, WT_CACHE_EVICT_CLEAN) && !F_ISSET(btree, WT_BTREE_IN_MEMORY) &&
+                      !modified) ||
           (F_ISSET(cache, WT_CACHE_EVICT_DIRTY) && modified) ||
-          (F_ISSET(cache, WT_CACHE_EVICT_UPDATES) && page->modify != NULL);
+          (F_ISSET(cache, WT_CACHE_EVICT_UPDATES) && !F_ISSET(btree, WT_BTREE_IN_MEMORY) &&
+            page->modify != NULL);
```
Comment (korteland) [inline src/evict/evict_lru.c:2107]: I'm not super familiar with eviction and this change is a little hard to follow - a comment in the code wouldn't be astray.

File: src/include/btree.h [hunk 3]
```diff
@@ -280,15 +280,16 @@ struct __wt_btree {
...
+#define WT_BTREE_IN_MEMORY 0x0020000u       /* In-memory object */
 #define WT_BTREE_LOGGED 0x0040000u          /* Commit-level durability without timestamps */
 #define WT_BTREE_NO_CHECKPOINT 0x0080000u   /* Disable checkpoints */
-#define WT_BTREE_OBSOLETE_PAGES 0x0100000u  /* Handle has obsolete pages */
-#define WT_BTREE_READONLY 0x0200000u        /* Handle is readonly */
-#define WT_BTREE_SALVAGE 0x0400000u         /* Handle is for salvage */
-#define WT_BTREE_SKIP_CKPT 0x0800000u       /* Handle skipped checkpoint */
-#define WT_BTREE_UPGRADE 0x1000000u         /* Handle is for upgrade */
-#define WT_BTREE_VERIFY 0x2000000u          /* Handle is for verify */
+#define WT_BTREE_NO_EVICT 0x0100000u        /* Cache-resident object */
```
Comment (korteland) [inline src/include/btree.h:286]: Reading through these, I can see myself forgetting the difference between `IN_MEMORY` and `NO_EVICT`, especially because in-memory would seem to imply no eviction. What about something that encodes the meaning a little better? `WT_BTREE_NO_CKPT` might be OK but it's already used. Perhaps `WT_BTREE_GARBAGE_COLLECTED`. (Alternatively, a comment above that contains more or less the same information as this PR's cover letter.)

**Review by korteland — APPROVED** [review]
Thanks - I have some comments but they're non-blocking. If we do indeed want to use this approach for solving the garbage collection problem then this looks like a good start.

File: src/include/btree.h (hunk 3 above)
Comment (quchenhao) [reply src/include/btree.h:286]: IN_MEMORY is different to no eviction. In memory database also do eviction but it doesn't remove it from memory. The same here.

Comment (korteland) [reply src/include/btree.h:286]: Right, but that's not obvious just from reading the definitions here. So either a comment or better names would be useful to me.

**Review by pmacko86 — APPROVED** [review]
LGTM, although I can't really review this as I don't understand this part of eviction. I am also unclear about when we need in-memory vs. no-evict; it would be nice if this was documented somewhere (if we don't have this already).

**Comment by quchenhao** [comment]
> LGTM, although I can't really review this as I don't understand this part of eviction. I am also unclear about when we need in-memory vs. no-evict; it would be nice if this was documented somewhere (if we don't have this already).

If the no-evict mechanism is enabled, a tree will not be considered for eviction. Therefore, it's in-memory footprint can only increase. But it will be visited by checkpoint and persisted to disk. The historical versions will also be moved to the history store by the checkpoint.

Oppositely, the new in-memory mechanism enables the tree for eviction and disables it for checkpoint. However, in eviction, we only reconciles the pages to replace the in-memory disk image to reduce its memory footprint but we will not write anything to disk. We use the same reconciliation mode for in-memory database to achieve that.

File: src/include/btree.h (hunk 3 above)
Comment (korteland) [reply src/include/btree.h]: Closing this out - the latest commits help with this.

**Comment by agorrod** [comment]
Thanks for including me, but I think my time has run out to consider the approach.

I trust you!
